File: OGLEngine/RemoteController.m
//
//  RemoteController.m
//  OGLEngine
//
//  Created by Maciej Chmielewski on 08.03.2016.
//  Copyright © 2016 MaciejCh. All rights reserved.
//

#import "RemoteController.h"
#import <SocketRocket/SRWebSocket.h>

@interface RemoteController () <SRWebSocketDelegate>

@property (nonatomic, strong) SRWebSocket *webSocketClient;
@property (nonatomic, strong) NSMutableArray<EventHandler> *eventHandlers;

@end

@implementation RemoteController

+ (instancetype)controller {
    static RemoteController *controller = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        controller = [[self alloc] init];
    });
    return controller;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        NSString *ipAddress = [NSString stringWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"IPAddress" ofType:@""]];
        self.webSocketClient = [[SRWebSocket alloc] initWithURL:[NSURL URLWithString:[NSString stringWithFormat:@"ws://%@:6001", ipAddress]]];
        [self.webSocketClient open];
        self.webSocketClient.delegate = self;
    }
    return self;
}

- (NSMutableArray<EventHandler> *)eventHandlers {
    if (!_eventHandlers) {
        _eventHandlers = [NSMutableArray new];
    }
    return _eventHandlers;
}

- (void)webSocket:(SRWebSocket *)webSocket didReceiveMessage:(NSString *)message {
    if (![message isKindOfClass:[NSString class]]) {
        return;
    }
    
    id eventSubject = nil;
    
    if ([message hasPrefix:@"m"]) {
        eventSubject = [RemoteMouse mouseWithMessage:message];
    } else if ([message hasPrefix:@"d"]) {
        eventSubject = [RemoteKey keyWithMessage:message];
    } else if ([message hasPrefix:@"u"]) {
        eventSubject = [RemoteKey keyWithMessage:message];
    }
    
    if (eventSubject) {
        for (EventHandler handler in self.eventHandlers) {
            handler(eventSubject);
        }
    }
    
}

- (void)addEventHandler:(EventHandler)eventHandler {
    [self.eventHandlers addObject:eventHandler];
}

- (void)removeEventHandler:(EventHandler)eventHandler {
    [self.eventHandlers removeObject:eventHandler];
}

@end

File: OGLEngine/RemoteController.swift
//
//  RemoteController.swift
//  OGLEngine
//
//  Created by Maciej Chmielewski on 27.03.2016.
//  Copyright © 2016 MaciejCh. All rights reserved.
//

import Foundation
import SwiftWebSocket

class RemoteController : NSObject {
    static let controller = RemoteController()
    
    var webSocketClient: WebSocket!
    var eventHandlers: [EventHandler] = []
    private var remotePropertiesCache: RemoteProperties!
    
    override init() {
        super.init()
        do {
            let ipAddress: String = try String(contentsOfFile:NSBundle.mainBundle().pathForResource("IPAddress", ofType: "")!)
            self.webSocketClient =  WebSocket("ws://\(ipAddress):6001")
            self.webSocketClient.event.message = {message in
                if let message = message as? String {
                    var eventSubject: Any? = nil
                    if message.hasPrefix("m") {
                        eventSubject = RemoteMouse.mouseWithMessage(message)
                    }
                    else if message.hasPrefix("d") {
                        eventSubject = RemoteKey.keyWithMessage(message)
                    }
                    else if message.hasPrefix("u") {
                        eventSubject = RemoteKey.keyWithMessage(message)
                    }
                    else if message.hasPrefix("p") {
                        let remoteProperty = RemoteProperties(jsonString: message.substringFromIndex(message.startIndex.advancedBy(2)))
                        self.remotePropertiesCache = remoteProperty
                        eventSubject = remoteProperty
                    }
                    
                    if let eventSubject = eventSubject {
                        for handler: EventHandler in self.eventHandlers {
                            handler(eventSubject: eventSubject)
                        }
                    }

                }
            }
        } catch _ {
        }
    }
    
    func integerNamed(name: String) -> Int {
        if remotePropertiesCache == nil {return 0}
        for property in remotePropertiesCache.properties {
            if property.name == name {
                if property.type == .Number {
                    return Int(property.value as! NSNumber)
                }
            }
        }
        return 0
    }
    
    func floatNamed(name: String) -> Float {
        if remotePropertiesCache == nil {return 0}
        for property in remotePropertiesCache.properties {
            if property.name == name {
                if property.type == .Number {
                    return Float(property.value as! NSNumber)
                }
            }
        }
        return 0
    }
    
    func addEventHandler(eventHandler: EventHandler) {
        self.eventHandlers.append(eventHandler)
    }
    
    func removeEventHandler(eventHandler: EventHandler) {
//        self.eventHandlers = self.eventHandlers.filter{$0 == eventHandler}
    }
}
